src/modules/vtol_att_control/standard.cpp

```diff
@@ -165,115 +164,99 @@ void Standard::update_vtol_state()
 	if (!_attc->is_fixed_wing_requested()) {
 
 		// the transition to fw mode switch is off
-		if (_vtol_schedule.flight_mode == MC_MODE) {
-			// in mc mode
-			_vtol_schedule.flight_mode = MC_MODE;
-			_mc_roll_weight = 1.0f;
-			_mc_pitch_weight = 1.0f;
-			_mc_yaw_weight = 1.0f;
-			_mc_throttle_weight = 1.0f;
-			_pusher_throttle = 0.0f;
-			_reverse_output = 0.0f;
 
-		} else if (_vtol_schedule.flight_mode == FW_MODE) {
+		if (_vtol_schedule.flight_mode == FW_MODE) {
 			// transition to mc mode
-			if (_vtol_vehicle_status->vtol_transition_failsafe == true) {
+			if (_vtol_vehicle_status->vtol_transition_failsafe) {
 				// Failsafe event, engage mc motors immediately
 				_vtol_schedule.flight_mode = MC_MODE;
 				_flag_enable_mc_motors = true;
-				_pusher_throttle = 0.0f;
-				_reverse_output = 0.0f;
-
 
 			} else {
 				// Regular backtransition
+				_vtol_schedule.transition_start = hrt_absolute_time();
 				_vtol_schedule.flight_mode = TRANSITION_TO_MC;
 				_flag_enable_mc_motors = true;
-				_vtol_schedule.transition_start = hrt_absolute_time();
-				_reverse_output = _params_standard.reverse_output;
-
+				_transition_achieved = false;
 			}
 
 		} else if (_vtol_schedule.flight_mode == TRANSITION_TO_FW) {
 			// failsafe back to mc mode
 			_vtol_schedule.flight_mode = MC_MODE;
-			_mc_roll_weight = 1.0f;
-			_mc_pitch_weight = 1.0f;
-			_mc_yaw_weight = 1.0f;
-			_mc_throttle_weight = 1.0f;
-			_pusher_throttle = 0.0f;
-			_reverse_output = 0.0f;
-
 
 		} else if (_vtol_schedule.flight_mode == TRANSITION_TO_MC) {
 			// transition to MC mode if transition time has passed or forward velocity drops below MPC cruise speed
-
 			const matrix::Dcmf R_to_body(matrix::Quatf(_v_att->q).inversed());
 			const matrix::Vector3f vel = R_to_body * matrix::Vector3f(_local_pos->vx, _local_pos->vy, _local_pos->vz);
 
-			float x_vel = vel(0);
+			const bool mc_speed_limit = _local_pos->v_xy_valid && (vel(0) <= _params_standard.mpc_xy_cruise);
 
-			if (hrt_elapsed_time(&_vtol_schedule.transition_start) >
-			    (_params_standard.back_trans_dur * 1000000.0f) ||
-			    (_local_pos->v_xy_valid && x_vel <= _params_standard.mpc_xy_cruise)) {
-				_vtol_schedule.flight_mode = MC_MODE;
+			const bool back_trans_timeout = (hrt_elapsed_time(&_vtol_schedule.transition_start) >
+							 (_params_standard.back_trans_dur * 1e6f));
+
+			if (back_trans_timeout || mc_speed_limit || can_transition_on_ground()) {
+				if (!_transition_achieved) {
+					// first time transition achieved, begin transition timer
+					_trans_finished_ts = hrt_absolute_time();
+				}
+
+				_transition_achieved = true;
+
+			} else {
+				_transition_achieved = false;
+				_trans_finished_ts = 0;
 			}
 
+			// all conditions must be continuously satisfied for > 1s to complete transition
```

Comment:
Did you encounter any false positive before? Only `mc_speed_limit` is susceptible to give a spike and usually `vel(.)` is quite smooth.

Reply:
I added this logic for forward transition (tested), and later decided to duplicate it for back (untested).